File: apps/api/src/rx_jev_api/routers/drugs.py
"""Drug search: name suggestions while typing, then the chosen name's RxCUI.

The RxCUI returned by `resolve` is the ingredient-set concept that `/api/labels/{rxcui}`
and its answers take.
"""

from typing import Annotated

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from rx_jev_api.clients.rxnorm import Ingredient
from rx_jev_api.deps import DrugNamesDep, RxNormDep
from rx_jev_api.suggest import suggest

router = APIRouter(prefix="/api/drugs", tags=["drugs"])

SUGGESTION_LIMIT = 10


class Suggestions(BaseModel):
    query: str
    # Names exactly as RxNorm spells them, including tall-man lettering like "metFORMIN".
    names: list[str]


class ResolvedDrug(BaseModel):
    query: str
    rxcui: str
    ingredients: list[Ingredient]


@router.get("/suggestions")
def read_suggestions(
    q: Annotated[str, Query(min_length=2, max_length=100)], names: DrugNamesDep
) -> Suggestions:
    return Suggestions(query=q, names=suggest(names, q, SUGGESTION_LIMIT))


@router.get("/resolve")
def resolve_drug(
    name: Annotated[str, Query(min_length=1, max_length=100)], rxnorm: RxNormDep
) -> ResolvedDrug:
    resolution = rxnorm.resolve(name)
    if resolution is None:
        raise HTTPException(status_code=404, detail=f"No drug found named {name!r}.")
    if resolution.rxcui is None:
        names = " and ".join(i.name for i in resolution.ingredients)
        raise HTTPException(
            status_code=404, detail=f"RxNorm has no single entry for {names} together."
        )
    return ResolvedDrug(
        query=resolution.query, rxcui=resolution.rxcui, ingredients=resolution.ingredients
    )

File: apps/worker/src/routes/drugs.ts
import type { ResolvedDrug, Suggestions } from '@rx-jev/contract'
import { suggest } from '@rx-jev/contract/suggest'
import { Hono } from 'hono'
import { z } from 'zod'
import type { AppEnv } from '../app.ts'
import { ProblemError, parseOr422 } from '../problems.ts'

// Drug search: name suggestions while typing, then the chosen name's RxCUI. The RxCUI
// returned by `resolve` is the ingredient-set concept the labels and answers routes take.

// Lengths count characters, as the Python API did.
function chars(min: number, max: number): z.ZodString {
  return z
    .string()
    .refine((s) => [...s].length >= min && [...s].length <= max, `${min} to ${max} characters`)
}

const SuggestionsQuery = z.object({ q: chars(2, 100) })
const ResolveQuery = z.object({ name: chars(1, 100) })

export const drugRoutes = new Hono<AppEnv>()
  .get('/suggestions', async (c) => {
    const { q } = parseOr422(SuggestionsQuery, c.req.query(), 'query')
    const names = await c.var.services.drugNames()
    return c.json<Suggestions>({ query: q, names: suggest(names, q) })
  })
  .get('/resolve', async (c) => {
    const { name } = parseOr422(ResolveQuery, c.req.query(), 'query')
    const resolution = await c.var.services.rxnorm.resolve(name)
    if (resolution === null) {
      throw new ProblemError(404, `No drug found named ${pythonRepr(name)}.`)
    }
    if (resolution.rxcui === null) {
      const names = resolution.ingredients.map((i) => i.name).join(' and ')
      throw new ProblemError(404, `RxNorm has no single entry for ${names} together.`)
    }
    return c.json<ResolvedDrug>({
      query: resolution.query,
      rxcui: resolution.rxcui,
      ingredients: resolution.ingredients,
    })
  })

// The detail quoted the name as Python's repr did: single quotes unless the name has one.
function pythonRepr(text: string): string {
  if (text.includes("'") && !text.includes('"')) return `"${text}"`
  return `'${text.replaceAll('\\', '\\\\').replaceAll("'", "\\'")}'`
}
